srml/fees/src/lib.rs

```diff
@@ -0,0 +1,132 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Handles all transaction fee related operations
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use srml_support::{dispatch::Result, Parameter, StorageMap, decl_event, decl_storage, decl_module};
+use runtime_primitives::traits::{As, Member, SimpleArithmetic, ChargeBytesFee, ChargeFee,
+	TransferAsset, CheckedAdd, CheckedSub, CheckedMul, Zero};
+use system;
+
+mod mock;
+mod tests;
+
+pub trait Trait: system::Trait {
+	/// The overarching event type.
+	type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+
+	/// The unit for fee amount
+	type Amount: Member + Parameter + SimpleArithmetic + Default + Copy + As<u64>;
+
+	/// A function does the asset transfer between accounts
+	type TransferAsset: TransferAsset<Self::AccountId, Amount = Self::Amount>;
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event<T>() = default;
+
+		fn on_finalise() {
+			let extrinsic_count = <system::Module<T>>::extrinsic_count();
+			(0..extrinsic_count).for_each(|index| {
+				// Deposit `Charged` event if some amount of fee charged.
+				let fee = <CurrentTransactionFee<T>>::take(index.clone());
+				if !fee.is_zero() {
+					Self::deposit_event(RawEvent::Charged(index.clone(), fee));
+				}
+			});
+		}
+	}
+}
+
+decl_event!(
+	pub enum Event<T> where <T as Trait>::Amount {
+		/// Fee charged (extrinsic_index, fee_amount)
+		Charged(u32, Amount),
+	}
+);
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Fees {
+		/// The fee to be paid for making a transaction; the base.
+		pub TransactionBaseFee get(transaction_base_fee) config(): T::Amount;
+		/// The fee to be paid for making a transaction; the per-byte portion.
+		pub TransactionByteFee get(transaction_byte_fee) config(): T::Amount;
+
+		/// The `extrinsic_index => accumulated_fees` map, containing records to
+		/// track the overall charged fees for each transaction.
+		///
+		/// All records should be removed at finalise stage.
+		CurrentTransactionFee get(current_transaction_fee): map u32 => T::Amount;
+	}
+}
+
+impl<T: Trait> ChargeBytesFee<T::AccountId> for Module<T> {
+	fn charge_base_bytes_fee(transactor: &T::AccountId, encoded_len: usize) -> Result {
+		let bytes_fee = match Self::transaction_byte_fee().checked_mul(
+			&<T::Amount as As<u64>>::sa(encoded_len as u64)
+		) {
```

Comment:
```suggestion
		).ok_or_else(|| "bytes fee overflow")?;
```

Is a little bit shorter :)